### PR TITLE
Improved Options for Termination Conditions

### PR DESCRIPTION
**Pull request by avik-pal**

### CODE REVIEW

File: src/termination.jl [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+            # Main Termination Criteria
+            objective <= criteria && return true
+
+            # Terminate if there has been no improvement for the last 30 steps
+            nstep += 1
+            push!(objective_values, objective)
+            last_30_values = objective_values[max(1, length(objective_values) - nstep):end]
+
+            objective <= 3 * criteria && nstep >= 30 &&
```
Comment (ChrisRackauckas) [inline src/termination.jl:71]: make that a parameter?

**Comment by ChrisRackauckas** [comment]
Yeah this looks mostly good. It would be good for that number 30 to be tweakable, and add some tests, and I think this is done.

File: src/solve.jl [hunk 2]
```diff
@@ -93,38 +93,46 @@ function DiffEqBase.__solve(prob::DiffEqBase.AbstractSteadyStateProblem,
...
-        u = sol.u[end][save_idxs]
+        u = sol.u[idx][save_idxs]
         du = du[save_idxs]
     else
-        u = sol.u[end]
+        u = sol.u[idx]
     end
 
-    if sol.retcode == ReturnCode.Terminated &&
-       (typeof(sol.u[end]) <: Array ? array_condition() : broadcast_condition())
```
Comment (avik-pal) [inline src/solve.jl:124]: @ChrisRackauckas, why was broadcast_condition not applicable for arrays?

Comment (ChrisRackauckas) [reply src/solve.jl:124]: it was just slow and slow to compile, it should be fine in modern Julia though. Some code is just old.

File: src/algorithms.jl [hunk 3]
```diff
@@ -32,14 +34,19 @@ sol = solve(prob,DynamicSS(CVODE_BDF()),dt=1.0)
...
 """
-struct DynamicSS{A, AT, RT, TS} <: SteadyStateDiffEqAlgorithm
+struct DynamicSS{A, AT, RT, TS, TC <: SteadyStateTerminationCriteria} <:
+       SteadyStateDiffEqAlgorithm
     alg::A
     abstol::AT
     reltol::RT
     tspan::TS
+
+    termination_condition::TC
```
Comment (ChrisRackauckas) [inline src/algorithms.jl:44]:
```suggestion
    termination_condition::TC
```

File: test/runtests.jl [hunk 4]
```diff
@@ -88,6 +88,27 @@ prob = SteadyStateProblem(ODEProblem(fiip, u0, tspan))
 sol2 = solve(prob, DynamicSS(Tsit5(), abstol = 1e-4))
 @test typeof(u0) == typeof(sol2.u)
 
+for mode in (:default, :norm, :rel, :rel_norm, :rel_safe, :rel_safe_best, :abs, :abs_norm,
```
Comment (ChrisRackauckas) [inline test/runtests.jl:91]: these should be enums? It seems that's what the EnumX would've been for?

File: src/termination.jl [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+            return _has_converged(get_du(integrator), integrator.u, $(mode_val), abstol,
+                                  reltol)
+        end
+    end
+end
+
+for mode in (:rel_safe, :rel_safe_best, :abs_safe, :abs_safe_best)
+    T = SteadyStateTerminationCriteria{mode}
+    mode_val = Val(mode)
+    @eval function _get_termination_condition(cond::$(T), storage)
```
Comment (ChrisRackauckas) [inline src/termination.jl:112]: do they need to be dispatches?

Comment (avik-pal) [reply src/termination.jl:112]: It feels wrong to have something that is known at compile time and make them runtime branches :sweat_smile:

File: test/runtests.jl (hunk 4 above)
Comment (avik-pal) [reply test/runtests.jl:91]: Done

File: src/termination.jl (hunk 5 above)
Comment (ChrisRackauckas) [reply src/termination.jl:145]: but is this improving runtime or just adding compile time?

Comment (avik-pal) [reply src/termination.jl:145]: This particular set of `@eval`s isn't adding any compile time. But the `_has_converged` ones add like `0.8ms` over `3.6ms` compilation time.

Comment (ChrisRackauckas) [reply src/termination.jl:145]: I still think it's a hyperoptimization to make this all dispatch, but we can keep it for now.

**Pull request merged**